gl.cpp

```diff
@@ -0,0 +1,164 @@
+	vector<int> textureIDs;	
+	for(int mat = 0; mat < materials.size(); mat++)
+	{
+		int tmp;
+		material m = materials[mat];
+		success &= loadTexture(m.map_Kd, tmp);
+		textureIDs.push_back(tmp);
+	}
+	validate(success, (char*)"Setup OpenGL Program");  
+
+
+	/***************************************************************************************
+	* VERTEX BUFFER OBJECT - Vertex data AND Attributes
+	***************************************************************************************/  
+	// Create VBO - Vertex Buffer Object - Which will hold the data (interleaved) for our vertices 
+	unsigned int VBO;
+	glGenBuffers(1, &VBO);
+	glBindBuffer(GL_ARRAY_BUFFER, VBO);
+	glBufferData(GL_ARRAY_BUFFER, vertexBufferNumBytes, vertexBufferData, GL_STATIC_DRAW);
+
+	
+	float data[] = { -0.5, -0.5, 1.0, 
+					 0.5,  -0.5, 1.0, 
+					 0.0,  0.0,  1.0};
+	unsigned int VBB; 
+	glGenBuffers(1, &VBB);
+	glBindBuffer(GL_ARRAY_BUFFER, VBB);
+	glBufferData(GL_ARRAY_BUFFER, 9*4, data, GL_STATIC_DRAW);
+		
+	/***************************************************************************
+	* ATTRIBUTE HANDLES - based on the way the shader is written; mapped to data 
+	***************************************************************************/      
+	int aPositionHandle = glGetAttribLocation(programID, "a_Position");
+	int aUVHandle = glGetAttribLocation(programID, "a_UV");
+
+
+	/***************************************************************************************
+	* UNIFORM HANDLES - kept to update uniforms; mapped for each frame (could be pre-cached)
+	***************************************************************************************/      
+	// Establish shader-specific variables
+	int uMatrixHandle = glGetUniformLocation(programID, "u_Matrix");
+	int uTextureHandle = glGetUniformLocation(programID, "u_Texture");
+	int uThresholdHandle = glGetUniformLocation(programID, "u_Threshold");
+
+	// MVP Data for transforming vertices
+	mat4 mvp;	  
+
+	// Camera data 
+	myCam.camX = myCam.camY = myCam.camZ = myCam.pitch = myCam.yaw = myCam.roll = 0.0;
+
+	// Main Loop 
+	int numDraw = vertexBuffer.size(); 
+	bool running = true;
+	printf("%s\n", glGetString(GL_VERSION));//_GetError());	
+	while(running)
+	{
+		// Update input
+		processUserInputs(running);      
+		{
+
+			// Setup Program, Attach appropriate buffer
+			glUseProgram(programID);
+			// Clear buffers, setup/use program 
+			glClearColor(0,0,0, 1.0);
+			glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
+
+			/***************************************************************************
+			* ATTRIBUTE HANDLES - based on the way the shader is written; mapped to data 
+			***************************************************************************/     
+			glBindBuffer(GL_ARRAY_BUFFER, VBO);
+			glEnableVertexAttribArray(aPositionHandle);
+			glVertexAttribPointer(aPositionHandle, 
+			3, GL_FLOAT, GL_FALSE, stride*sizeof(float), (void*)0);
+			glEnableVertexAttribArray(aUVHandle);
```

Comment:
We also don't need to find the attribute locations each time because our shaders don't change